tutorials/haskell/snap-file-uploads/tutorial.md

```diff
@@ -0,0 +1,324 @@
+---
+title: File uploads in (a) Snap
+published: 2016-07-12
+ghc: 7.10.3
+lts: 6.5
+libraries: snap-core-0.9.8.0 snap-server-0.9.5.1
+language: Haskell
+author-name: Juan Pedro Villa Isaza
+---
+
+> “I was also sad, and thinking,  
+> When one day I saw you winking,  
+> And I heard you sniffle-snuffle,  
+> And I saw your feathers ruffle.
+
+We recently had to add file uploads to a
+[Haskell](https://www.haskell.org/) web application built with the
+[Snap](http://snapframework.com/) framework. File uploads in Snap (that
+is,
+[`Snap.Util.FileUploads`](https://hackage.haskell.org/package/snap-core-0.9.8.0/docs/Snap-Util-FileUploads.html))
+have good documentation, but we thought it would be useful to write a
+tutorial about how to implement a basic file uploader. The idea is to
+write an application with an HTML form to upload multiple Haskell files
+(`.hs` or `.lhs`) that simply displays the contents of the uploaded
+files.
+
+This blog post was generated from a literate Haskell file that you can
+find in the
+[turbulent-sniffle](https://github.com/stackbuilders/turbulent-sniffle)
+repository. All the code was tested with [LTS Haskell
+6.5](https://www.stackage.org/lts-6.5), that is, GHC 7.10.3,
+[snap-core
+0.9.8.0](https://hackage.haskell.org/package/snap-core-0.9.8.0), and
+[snap-server
+0.9.5.1](https://hackage.haskell.org/package/snap-server-0.9.5.1).
+
+The first thing we'll do is create an `index.html` file and add an HTML
+form to upload multiple files:
+
+``` {.html}
+<form action="/" method="post" enctype="multipart/form-data">
+  <input type="file" name="files" multiple>
+  <button type="submit">Submit</button>
+</form>
+```
+
+This is just an HTML form with a file input that allows multiple files.
+We could make the input required, but we'll use Haskell to handle
+everything.
+
+In order to see exactly what we need for adding file uploads to an
+existing Haskell application, let's create a Cabal file. In this case,
+we'll call it `turbulent-sniffle.cabal`:
+
+    name: turbulent-sniffle
+    version: 0.1.0
+    build-type: Simple
+    cabal-version: >= 1.22
+
+We can now add the `index.html` file as a data file:
+
+    data-files: index.html
+
+And create an executable component:
+
+    executable turbulent-sniffle
+      main-is: Main.lhs
+      other-modules: Paths_turbulent_sniffle
+      ghc-options: -Wall
+      build-depends:
```

Comment:
@seuros The imports below specify what we're using from each library. snap-server installs the rest, but we have to include them here so they're not hidden.